### ISSUE TITLE
Assertion in  file SelectionDAG.cpp:1454 when trying to lower @llvm.masked.scatter with opaque pointers on X86 

### ISSUE THREAD

**Issue by fhahn** [llvm/llvm-project#55021]
`llc` triggers the following assertion for the code below: https://llvm.godbolt.org/z/jME9oYrbh

```console
SelectionDAG.cpp:1454: llvm::SDValue llvm::SelectionDAG::getConstant(uint64_t, const llvm::SDLoc&, llvm::EVT, bool, bool): Assertion `(EltVT.getSizeInBits() >= 64 || (uint64_t)((int64_t)Val >> EltVT.getSizeInBits()) + 1 < 2) && "getConstant with a uint64_t value that doesn't fit in the type!"' failed
```

```llvm
define void @crash(ptr %result, <4 x i64>* %idx.ptr, <4 x i1>* %mask.ptr) #0 {
entry:
  %idx = load <4 x i64>, <4 x i64>* %idx.ptr
  %mask = load <4 x i1>, <4 x i1>* %mask.ptr
  %0 = getelementptr inbounds [512 x double], ptr %result, <4 x i64> %idx
  call void @llvm.masked.scatter.v4f64.v4p0(<4 x double> zeroinitializer, <4 x ptr> %0, i32 0, <4 x i1> %mask)
  ret void
}

; Function Attrs: nocallback nofree nosync nounwind willreturn writeonly
declare void @llvm.masked.scatter.v4f64.v4p0(<4 x double>, <4 x ptr>, i32 immarg, <4 x i1>) #1

attributes #0 = { "target-cpu"="skylake-avx512" }
```

**Comment by phoebewang** [comment]
I'm not familiar with opaque pointer, but the IR cannot be compiled with typed pointer if we replace `ptr` with `[512 x double]*`.
It only works in below cases: https://llvm.godbolt.org/z/6hxTfcPrY, https://godbolt.org/z/YGWYjeoc3
@paulwalker-arm Is it somehow related to your fix https://github.com/llvm/llvm-project/commit/0c44115e5120167fc573e36dd878f4f95f5d63e6 ?

**Comment by paulwalker-arm** [comment]
[0c44115](https://github.com/llvm/llvm-project/commit/0c44115e5120167fc573e36dd878f4f95f5d63e6) is likely related but not the cause. My patch fixes an issues where invalid masked gather/scatter ISD nodes are created that resulted in the "Scale should be a constant power of 2" assert firing for many SVE workloads after opaque pointers was enabled.  The link to opaque pointer is because previously the IR contained pointer bitcasts that meant it failed the "GetElementPointer" test within getUniformBase().  With opaque pointers such bitcasts are no longer necessary and thus don't exist in the IR.  The net result being we now start creating scaled masked gather/scatter nodes when previously we did not.

This to me looks like another instance of this, but this time I think it's the way Scale is calculated? We've always used "GEP->getResultElementType()", but that assumes this will be the same type as used by the gather/scatter intrinsics, which in the opaque pointer world is not true.

**Comment by alexey-bataev** [comment]
Just for the reference, looks like caused by https://reviews.llvm.org/rG87a0b1bd233a3680c32a6e17acf147d0fe90c9e9

**Comment by nikic** [comment]
It's not entirely clear to me what the contract for MSCATTER/MGATHER nodes is. Should we only filter out scales larger than supported by target constant (to avoid this assertion failure), or should we add a check that the scale matches the element size of the gather/scatter itself?

**Comment by nikic** [comment]
https://github.com/llvm/llvm-project/commit/0c44115e5120167fc573e36dd878f4f95f5d63e6 went with the interpretation that scales not matching the element type are allowed, but I'm not sure if that's really desired -- it seems to effectively more the check into the SVE implementation, so maybe we want it to be part of DAG builder instead?

**Comment by paulwalker-arm** [comment]
The MSCATTER/MGATHER ISD nodes have two ways to represent similar logic.  For SVE we added an enum to represent scaled vs unscaled (and this mandated that scaling was a function of the MemVT element type), but before/whilst adding that support the nodes gained an explicit Scale parameter.  I've got a set of patches to remove the duplication, plus fix a bunch of bugs I've discovered along the way.  I figured that because somebody added the explicit parameter and it provides more flexibility than the enum we should generalise to that.  I do agree this "flexibility" means we potentially end up with duplicate legalisation code, but that's something that can be resolved whilst still keeping the flexibility, through common expand code.  That said, if the consensus is to limit this flexibility then I'm fine with that also.
